torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -534,7 +534,13 @@ def _root_pre_forward(
         if handle:
             should_cast_buffers_to_full_prec = handle._force_full_precision
         else:
-            should_cast_buffers_to_full_prec = True
+            # If the root has no handle (no managed parameters), then we fall
+            # back to checking if any child wants to force full precision as a
+            # workaround
```

Comment:
This is a hack, but the code is somewhat beyond saving.

The root FSDP module is responsible for this cast logic, but each FSDP submodule could have different values for `_force_full_precision`, in which case the result is wrong. We could hard error here in that case.